src/gd_avif.c

```diff
@@ -0,0 +1,646 @@
+/**
+ * File: AVIF IO
+ *
+ * Read and write AVIF images using libavif (https://github.com/AOMediaCodec/libavif) .
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <limits.h>
+#include <math.h>
+
+#include "gd.h"
+#include "gd_errors.h"
+#include "gdhelpers.h"
+#include "gd_intern.h"
+
+#ifdef HAVE_LIBAVIF
+#include <avif/avif.h>
+
+// TODO: more error checking, everywhere
+// TODO: make sure all indents are tabs, not spaces
+
+/*
+	Define defaults for encoding images:
+		DEFAULT_CHROMA_SUBSAMPLING: 4:2:0 is commonly used for Chroma subsampling.
+		DEFAULT_MIN_QUANTIZER, DEFAULT_MAX_QUANTIZER: 
+			We need more testing to really know what quantizer settings are optimal,
+			but teams at Google have been using maximum=30 as a starting point.
+		DEFAULT_QUALITY: following gd conventions, -1 indicates the default.
+		DEFAULT_SPEED: AVIF_SPEED_DEFAULT is -1. This simply tells the AVIF encoder to use the default speed.
+*/
+
+#define DEFAULT_CHROMA_SUBSAMPLING AVIF_PIXEL_FORMAT_YUV420
+#define DEFAULT_QUANTIZER 30
+#define DEFAULT_QUALITY -1
+#define DEFAULT_SPEED AVIF_SPEED_DEFAULT
+
+// This initial size for the gdIOCtx is standard among GD image conversion functions.
+#define NEW_DYNAMIC_CTX_SIZE 2048
+
+// Our quality param ranges from 0 to 100.
+// To calculate quality, we convert from AVIF's quantizer scale, which runs from 63 to 0.
+#define MAX_QUALITY 100
+
+// These constants are for computing the number of tiles and threads to use during encoding.
+// Maximum threads are from libavif/contrib/gkd-pixbuf/loader.c.
+#define MIN_TILE_AREA (512 * 512)
+#define MAX_TILES 8
+#define MAX_THREADS 64
+
+
+/*** Helper functions ***/
+
+/* Convert the quality param we expose to the quantity params used by libavif.
+	 The *Quantizer* params values can range from 0 to 63, with 0 = highest quality and 63 = worst.
+	 We make the scale 0-100, and we reverse this, so that 0 = worst quality and 100 = highest.
+
+	 Values below 0 are set to 0, and values below MAX_QUALITY are set to MAX_QUALITY.
+*/
+static int quality2Quantizer(int quality) {
+	int clampedQuality = CLAMP(quality, 0, MAX_QUALITY);
+
+	float scaleFactor = (float) AVIF_QUANTIZER_WORST_QUALITY / (float) MAX_QUALITY;
+
+	return round(scaleFactor * (MAX_QUALITY - clampedQuality));
+}
+
+/*
+	From gd_png.c:
+		convert the 7-bit alpha channel to an 8-bit alpha channel.
+		We do a little bit-flipping magic, repeating the MSB
+		as the LSB, to ensure that 0 maps to 0 and
+		127 maps to 255. We also have to invert to match
+		PNG's convention in which 255 is opaque. 
+*/
+static uint8_t convertTo8BitAlpha(uint8_t originalAlpha) {
+	return originalAlpha == 127 ? 
+				0 : 
+				255 - ((originalAlpha << 1) + (originalAlpha >> 6));
+}
+
+/*
+	 As of February 2021, this algorithm reflects the latest research on how many tiles
+	 and threads to include for a given image size.
+	 This is subject to change as research continues.
```

Comment:
This algorithms could be much simpler. There is no need to calculate tiles first. Also the floor(log2()) can just be a small lookup table.

      encoder->tileColsLog2 = floor_log2(width / width_tile_size) / 2;
      encoder->tileRowsLog2 = floor_log2(height / height_tile_size) / 2;

Reply:
lets avoid premature optimization with building lookup tables.  i suspect 2 x (floor+log2) will be system noise compared to the AV1 codec in libavif itself.

lets merge a working & straight forward implementation, and if you want to profile the code to find hot spots, we can do it as a follow up.  but i strongly suspect GD is never the hot spot.

Reply:
Agreeing with that - this operation is only done once per encoding, and it will take a fraction of a second. We'll spend much more time in the AVIF encoding process, which can easily take 1-60 seconds.

Reply:
Also, if it helps, I'm changing this to the version @wantehchang proposed, using `log2()` directly.

Reply:
FloorLog2 is a common operation in this kind of code. libgd may already have an optimized implementation that goes by some other name such as "count leading zeros" or "bit scan reverse". See https://en.wikipedia.org/wiki/Binary_logarithm#Integer_rounding.